### PR TITLE
exchanges/account: add UpdatedAt field to Balance and ProtectedBalance

### PR DESCRIPTION
**Pull request by junnplus**
# PR Description

Adding a timestamp to balance updates and improving the handling of balance changes. This change is made to retrieve updated account balance info via WebSocket in the future.

### Enhancements to balance handling:

* `exchanges/account/account_types.go`: Added an `UpdatedAt` field to both the `Balance` and `ProtectedBalance` structs to track the last update time. 

* `exchanges/account/account.go`: Modified the `load` method in the `ProtectedBalance` struct to return a boolean indicating whether the balance was updated, and added logic to check the `UpdatedAt` field before updating the balance.

### Test updates:

* `exchanges/account/account_test.go`: Updated the `TestBalanceInternalLoad` tests to check the `UpdatedAt` field and validate the new behavior of the `load` method.

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [x] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/account/account.go [hunk 1]
```diff
@@ -268,22 +268,34 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
...
+func (b *ProtectedBalance) load(change *Balance) bool {
 	b.m.Lock()
 	defer b.m.Unlock()
+	if change == nil {
+		return false
+	}
+	if b.updatedAt.After(change.UpdatedAt) {
+		return false
+	}
+	if !b.updatedAt.IsZero() && b.updatedAt == change.UpdatedAt {
```
Comment (shazbert) [inline exchanges/account/account.go:280]: Can use equal method.
```suggestion
	if !b.updatedAt.IsZero() && b.updatedAt.Equal(change.UpdatedAt) {
```

Also suggestion we can probably merge the above if case with `b.updatedAt.After(change.UpdatedAt)`

```go
if !b.updatedAt.IsZero() {
		if b.updatedAt.After(change.UpdatedAt) {
			return false
		}
		if b.updatedAt.Equal(change.UpdatedAt) {
		return false
	}
}
```

File: exchanges/account/account.go [hunk 2]
```diff
@@ -268,22 +268,34 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
...
+		return false
+	}
 	if b.total == change.Total &&
 		b.hold == change.Hold &&
 		b.free == change.Free &&
 		b.availableWithoutBorrow == change.AvailableWithoutBorrow &&
-		b.borrowed == change.Borrowed {
-		return
+		b.borrowed == change.Borrowed &&
+		b.updatedAt == change.UpdatedAt {
```
Comment (shazbert) [inline exchanges/account/account.go:288]: Use Equal method

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for this PR. This update doesn't enforce balances to have UpdateAt set specifically at the exchange level but it is a good thing to have once websocket balance updates are handled through this package correctly. So I am not against this change. But here are some simple nits.

File: exchanges/account/account.go [hunk 3]
```diff
@@ -268,22 +268,34 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
 
 // load checks to see if there is a change from incoming balance, if there is a
 // change it will change then alert external routines.
-func (b *ProtectedBalance) load(change Balance) {
+func (b *ProtectedBalance) load(change *Balance) bool {
```
[the remark this answers is not in the document]
Comment (junnplus) [reply exchanges/account/account.go]: That makes sense. I updated the implementation accordingly.

File: exchanges/account/account.go (hunk 1 above)
Comment (junnplus) [reply exchanges/account/account.go]: I merged the above.
```
	if !b.updatedAt.IsZero() && !b.updatedAt.Before(change.UpdatedAt) {
		return errOutOfSequence
	}
```

File: exchanges/account/account.go [hunk 4]
```diff
@@ -268,22 +275,31 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
 
 // load checks to see if there is a change from incoming balance, if there is a
 // change it will change then alert external routines.
-func (b *ProtectedBalance) load(change Balance) {
+func (b *ProtectedBalance) load(change *Balance) error {
 	b.m.Lock()
 	defer b.m.Unlock()
+	if change == nil {
+		return fmt.Errorf("%w for '%T'", common.ErrNilPointer, change)
+	}
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/account/account.go:283]: Also could you please have a test case with a nil balance for coverage.

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks, last nits then looks good to go. 🚀

**Review by shazbert — APPROVED** [review]
Changes approved thanks!

File: exchanges/account/account.go [hunk 5]
```diff
@@ -268,22 +275,31 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
...
+func (b *ProtectedBalance) load(change *Balance) error {
+	if change == nil {
+		return fmt.Errorf("%w for '%T'", common.ErrNilPointer, change)
+	}
 	b.m.Lock()
 	defer b.m.Unlock()
+	if !b.updatedAt.IsZero() && !b.updatedAt.Before(change.UpdatedAt) {
+		return errOutOfSequence
+	}
 	if b.total == change.Total &&
```
Comment (gloriousCode) [inline exchanges/account/account.go:287]: While I appreciate the added field, looking across all usages of `Balance`, no exchange populates the field `UpdatedAt`.
You could do something as simple as
`if change.UpdatedAt.IsZero() { change.UpdatedAt = time.Now() }`
Or you could populate it at all the sources

File: exchanges/account/account_types.go [hunk 6]
```diff
@@ -59,6 +60,7 @@ type Balance struct {
 	Free                   float64
 	AvailableWithoutBorrow float64
 	Borrowed               float64
+	UpdatedAt              time.Time
```
Comment (gloriousCode) [inline exchanges/account/account_types.go:63]: Main thing I'd like to request of you is to update the rpc server definitions to include this.
in `/gctrpc/rpc.proto` update `message AccountCurrencyInfo {` to have `UpdatedAt`, regenerate proto definitions via `/gctrpc/buf generate`

File: exchanges/account/account.go [hunk 7]
```diff
@@ -103,7 +104,7 @@ func GetHoldings(exch string, creds *Credentials, assetType asset.Item) (Holding
 		return Holdings{}, fmt.Errorf("%s %s %s %w %w", exch, creds, assetType, errNoCredentialBalances, ErrExchangeHoldingsNotFound)
```
Comment (gloriousCode) [inline exchanges/account/account.go:104]: If you could also update `GetHoldings` to actually return the `UpdatedAt` value, it would be handy to receive

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thank you for your contribution! I understand that this is trying to be forward looking, its just that nothing supports the field you've added yet and this could be added later with that support.
I've added suggestions so that the field actually gets utilised within this PR. Let me know if you're up for them. I can still approve this as-is despite me asking for more.

File: exchanges/account/account.go (hunk 5 above)
Comment (junnplus) [reply exchanges/account/account.go:291]: Fixed it by https://github.com/thrasher-corp/gocryptotrader/pull/1827/commits/fa0651d8341cb22f53ff8154d1e991ce6d5b8732

File: exchanges/account/account.go [hunk 8]
```diff
@@ -268,22 +279,31 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
...
 // load checks to see if there is a change from incoming balance, if there is a
 // change it will change then alert external routines.
-func (b *ProtectedBalance) load(change Balance) {
+func (b *ProtectedBalance) load(change *Balance) error {
+	if change == nil {
+		return fmt.Errorf("%w for '%T'", common.ErrNilPointer, change)
+	}
 	b.m.Lock()
 	defer b.m.Unlock()
+	if !b.updatedAt.IsZero() && !b.updatedAt.Before(change.UpdatedAt) {
```
Comment (shazbert) [inline exchanges/account/account.go:288]: Now that the time is assigned in Update method. Could we please force storage, so if `!b.updatedAt.IsZero()` we then return an error. 🙏 And update the test.

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for the changes. One minor request regarding new update.

File: exchanges/account/account.go [hunk 9]
```diff
@@ -268,22 +280,34 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
...
-func (b *ProtectedBalance) load(change Balance) {
+func (b *ProtectedBalance) load(change *Balance) error {
+	if change == nil {
+		return fmt.Errorf("%w for '%T'", common.ErrNilPointer, change)
+	}
 	b.m.Lock()
 	defer b.m.Unlock()
+	if change.UpdatedAt.IsZero() {
+		return errUpdatedAtIsZero
+	}
```
Comment (shazbert) [inline exchanges/account/account.go:291]:
```suggestion
        if change.UpdatedAt.IsZero() {
		return errUpdatedAtIsZero
	}
	b.m.Lock()
	defer b.m.Unlock()
```

**Review by shazbert — APPROVED** [review]
One suggestion, but this is tACK. thanks @junnplus 💃

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for making those changes

(balance is fake)
```
go run . getaccountinfo --exchange="btc markets" --asset="spot"
{
 "exchange": "btc markets",
 "accounts": [
  {
   "id": "Key:[1337...] SubAccount:[] ClientID:[]",
   "currencies": [
    {
     "currency": "BTC",
     "total_value": 1337,
     "free": 1337,
     "updated_at": "2025-03-11 16:07:04 AEDT"
    }
   ]
  }
 ]
}%             
```

**Review by thrasher- — APPROVED** [review]
Thanks for making those changes!

File: exchanges/account/account_test.go [hunk 10]
```diff
@@ -292,8 +296,20 @@ func TestBalanceInternalWait(t *testing.T) {
...
-	bi.load(Balance{Total: 1, Hold: 2, Free: 3, AvailableWithoutBorrow: 4, Borrowed: 5})
+	err := bi.load(nil)
+	assert.ErrorIs(t, err, common.ErrNilPointer, "should error nil pointer correctly")
+
+	err = bi.load(&Balance{Total: 1, Hold: 2, Free: 3, AvailableWithoutBorrow: 4, Borrowed: 5})
+	assert.ErrorIs(t, err, errUpdatedAtIsZero, "should have not been loaded")
+
+	now := time.Now()
+	err = bi.load(&Balance{UpdatedAt: now, Total: 1, Hold: 2, Free: 3, AvailableWithoutBorrow: 4, Borrowed: 5})
+	assert.NoError(t, err, "should have been loaded")
```
Comment (gbjk) [inline exchanges/account/account_test.go:307]: 🦟 Nit: Since this is a unit test, and this is an NoError on the main function called, the comment can be omitted.
If it's not omitted, it should say "Must not error when X" if X is a regression case or it's not clear what obsticle has been cleared. In this case, UpdatedAt was the last error, we've now provided it, so it's obvious why it should now pass.
🚧 It also feels like this is a pre-requisit for the following, so needs to be require.
```suggestion
	require.NoError(t, err)
```

File: exchanges/account/account_test.go [hunk 11]
```diff
@@ -292,8 +296,20 @@ func TestBalanceInternalWait(t *testing.T) {
...
+
 	bi.m.Lock()
+	if bi.updatedAt.IsZero() {
+		t.Fatal("unexpected value")
+	}
```
Comment (gbjk) [inline exchanges/account/account_test.go:312]: 🚧 Can we check equality please?
```suggestion
    	assert.Equal(t, now, bi.updatedAt, "updatedAt should contain the correct value")
```
Or alternatively:
```
    	assert.False(t, bi.updatedAt.IsZero(), "updatedAt should not be IsZero")
```
Or finally:
```
    	assert.NotEmpty(t, bi.updatedAt, "updatedAt should contain a value")
```

File: exchanges/account/account_test.go [hunk 12]
```diff
@@ -314,6 +330,12 @@ func TestBalanceInternalLoad(t *testing.T) {
 	if bi.GetFree() != 3 {
 		t.Fatal("unexpected value")
 	}
+
+	err = bi.load(&Balance{UpdatedAt: now, Total: 2, Hold: 3, Free: 4, AvailableWithoutBorrow: 5, Borrowed: 6})
+	assert.Error(t, err, "should have not been loaded")
```
Comment (gbjk) [inline exchanges/account/account_test.go:335]: 🚧 Can we please check the error value?
```suggestion
	assert.ErrorIs(t, errOutOfSequence, err, "should error correctly with same UpdatedAt")
```

File: exchanges/account/account.go [hunk 13]
```diff
@@ -31,6 +31,8 @@ var (
 	errBalanceIsNil                 = errors.New("balance is nil")
 	errNoCredentialBalances         = errors.New("no balances associated with credentials")
 	errCredentialsAreNil            = errors.New("credentials are nil")
+	errOutOfSequence                = errors.New("out of sequence")
+	errUpdatedAtIsZero              = errors.New("updated at is zero")
```
Comment (gbjk) [inline exchanges/account/account.go:35]: 🦟 Nitpick: I think that this needs to be `updatedAt` 🙂 
Also can reword to make it clearer this error is an error, since it's going down the pb wire.
```suggestion
	errUpdatedAtIsZero              = errors.New("updatedAt may not be zero")
```

**Review by gbjk — CHANGES_REQUESTED** [review]
Great work, Thank you 🙏 🎉 

Just a few minor change requests 🚧 from me.
A few nitpicks 🦟  which you can ignore or action at your discretion.

File: exchanges/account/account.go [hunk 14]
```diff
@@ -268,22 +280,34 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
 
 // load checks to see if there is a change from incoming balance, if there is a
 // change it will change then alert external routines.
-func (b *ProtectedBalance) load(change Balance) {
+func (b *ProtectedBalance) load(change *Balance) error {
+	if change == nil {
+		return fmt.Errorf("%w for '%T'", common.ErrNilPointer, change)
+	}
```
[the remark this answers is not in the document]
Comment (junnplus) [reply exchanges/account/account.go:286]: NilGuard looks solid! 👏

File: exchanges/account/account.go [hunk 15]
```diff
@@ -268,22 +280,34 @@ func (s *Service) Update(incoming *Holdings, creds *Credentials) error {
...
+		return errOutOfSequence
+	}
 	if b.total == change.Total &&
 		b.hold == change.Hold &&
 		b.free == change.Free &&
 		b.availableWithoutBorrow == change.AvailableWithoutBorrow &&
-		b.borrowed == change.Borrowed {
-		return
+		b.borrowed == change.Borrowed &&
+		b.updatedAt.Equal(change.UpdatedAt) {
```
[the remark this answers is not in the document]
Comment (junnplus) [reply exchanges/account/account.go:300]: We can add it later if necessary.

File: exchanges/account/account.go [hunk 16]
```diff
@@ -32,7 +32,8 @@ var (
 	errNoCredentialBalances         = errors.New("no balances associated with credentials")
 	errCredentialsAreNil            = errors.New("credentials are nil")
 	errOutOfSequence                = errors.New("out of sequence")
-	errUpdatedAtIsZero              = errors.New("updated at is zero")
+	errUpdatedAtIsZero              = errors.New("updatedAt may not be zero")
+	errLoadingBalance               = errors.New("cannot loading balance")
```
Comment (gbjk) [inline exchanges/account/account.go:36]: 🚧  Just a minor typo here:
```suggestion
	errLoadingBalance               = errors.New("cannot load balance")
```
Alternative: `error loading balance`

File: exchanges/account/account_test.go [hunk 17]
```diff
@@ -300,42 +300,28 @@ func TestBalanceInternalLoad(t *testing.T) {
...
 	err = bi.load(&Balance{Total: 1, Hold: 2, Free: 3, AvailableWithoutBorrow: 4, Borrowed: 5})
-	assert.ErrorIs(t, err, errUpdatedAtIsZero, "should have not been loaded")
+	assert.ErrorIs(t, err, errUpdatedAtIsZero, "should error correctly when updatedAt is not set")
 
 	now := time.Now()
 	err = bi.load(&Balance{UpdatedAt: now, Total: 1, Hold: 2, Free: 3, AvailableWithoutBorrow: 4, Borrowed: 5})
-	assert.NoError(t, err, "should have been loaded")
+	require.NoError(t, err)
 
 	bi.m.Lock()
```
Comment (gbjk) [inline exchanges/account/account_test.go:309]: 🥃 Aside: You didn't introduce this, but I just realised that this is a single test accessing `bi` directly.
We really don't need to lock it 🤷

File: exchanges/account/account_test.go [hunk 18]
```diff
@@ -300,42 +300,28 @@ func TestBalanceInternalLoad(t *testing.T) {
...
 	bi.m.Unlock()
 
-	if bi.GetFree() != 3 {
-		t.Fatal("unexpected value")
-	}
+	assert.Equal(t, 3.0, bi.GetFree())
 
 	err = bi.load(&Balance{UpdatedAt: now, Total: 2, Hold: 3, Free: 4, AvailableWithoutBorrow: 5, Borrowed: 6})
-	assert.Error(t, err, "should have not been loaded")
+	assert.ErrorIs(t, err, errOutOfSequence, err, "should error correctly with same UpdatedAt")
```
Comment (gbjk) [inline exchanges/account/account_test.go:321]: 🚧 
```suggestion
	assert.ErrorIs(t, err, errOutOfSequence, "should error correctly with same UpdatedAt")
```

**Review by gbjk — CHANGES_REQUESTED** [review]
Thanks for those changes 👍 
Just a few quick typos, and one issue with the func signature for `assert.ErrorIs`

File: exchanges/account/account_test.go (hunk 17 above)
Comment (junnplus) [reply exchanges/account/account_test.go:309]: let’s keep it as is for now.

**Review by gbjk — APPROVED** [review]
LGTM 👍
